core/src/main/java/com/oath/oak/EntrySet.java

```diff
@@ -473,365 +432,257 @@ boolean casNextEntryIndex(int ei, int nextOld, int nextNew) {
 
     /**
-     * getKeyBlockIDFromIntArray extracts keys BlockID from the translation of the key reference (long)
-     * into array of two integers. This is done in order to avoid more int[] objects.
+     * Updates the key portion of the entry context inside {@code ctx} from an entry given by entry index {@code ei}.
+     * The entry index (ei) is stored in the context so it can  be used later by other methods without passing the
+     * entry index explicitly as a parameter.
+     * Specifically, this call might be followed up by {@code readValue(ctx)}.
      *
-     * Always use >>> (unsigned right shift) operator. It always fills 0 irrespective
-     * of the sign of the number.
+     * @param ctx the context that will be updated and follows the operation with this key
+     * @param ei  the entry index to look up
+     * @return    true if the entry index has a valid key reference
      */
-    static private int getKeyBlockIDFromIntArray(int[] long2IntTranslated) {
-        return long2IntTranslated[BLOCK_ID_LENGTH_ARRAY_INDEX] >>> KEY_BLOCK_SHIFT;
+    boolean readKey(ThreadContext ctx, int ei) {
+        ctx.entryIndex = ei;
+        return readKey(ctx.key, ei);
     }
 
     /**
-     * getValueBlockIDFromIntArray extracts value BlockID from the translation of the value reference (long)
-     * into array of two integers. This is done in order to avoid more int[] objects.
+     * Updates the value portion of the entry context inside {@code ctx} that matches the look-up context key.
+     * This includes both the value itself, and the value's state.
+     * Thus, {@code readKey(ctx, ei)} should be called prior to this method with the same {@code ctx} instance.
      *
-     * Always use >>> (unsigned right shift) operator. It always fills 0 irrespective
-     * of the sign of the number.
+     * @param ctx the context that was initiated by {@code readKey(ctx, ei)}
      */
-    static private int getValueBlockIDFromIntArray(int[] long2IntTranslated) {
-        return long2IntTranslated[BLOCK_ID_LENGTH_ARRAY_INDEX] >>> VALUE_BLOCK_SHIFT;
+    boolean readValue(ThreadContext ctx) {
+        boolean isValid = readValue(ctx.value, ctx.entryIndex);
+        ctx.valueState = getValueState(ctx.value);
+        return isValid;
     }
 
     /**
-     * getPositionFromIntArray extracts in Block position from the translation of the off-heap
-     * reference (long) into array of two integers. This is done in order to limit int[] objects.
-     */
-    static private int getPositionFromIntArray(int[] long2IntTranslated) {
-        return long2IntTranslated[POSITION_ARRAY_INDEX];
-    }
+     * Find the state of a the value that is pointed by {@code value}.
+     * Thus, {@code readValue(value, ei)} should be called prior to this method with the same {@code value} instance.
+     *
+     * @param value a buffer object that contains the value buffer
+     */
+    ValueState getValueState(ValueBuffer value) {
```

Comment:
Initially, I considered making this method public. Letting the `EntrySet` user decide when she wants to read the value's state, after calling `readValue()`. But eventually, I never used it anywhere else.
Technically, it CAN be used outside of this method, e.g., for reading the state again to check if it might changed. So I left it as is.